### ISSUE TITLE
client/asset: handle gap limits

### ISSUE THREAD

**Issue by buck54321** [decred/dcrdex#215]
Right now, the clients are creating new addresses with reckless disregard for the wallet's gap limit. This must be fixed. I can think of a few solutions, listed from easiest to hardest. 

1. Never create new addresses. Maybe allow the client to trigger an address cycle.
2. Don't create new addresses until old ones are seen.
3. Employ an algorithm to check `listreceivedbyaddress` for unused addresses before asking for new ones. 
4. Use persistent storage to track addresses across sessions and handle address creation based on that record. This would require us to provide the `ExchangeWallet` with probably a `.dexc` subdirectory to use, and the ExchangeWallet to maintain a database.

**Comment by chappjc** [comment]
This was at least partially resolved with DCR by using the "ignore" gap policy limit: https://github.com/decred/dcrdex/pull/257.  For BTC, the gap limit is 1000.

**Comment by buck54321** [comment]
I'm okay closing this if you are, @chappjc. It doesn't appear that BTC will ever stop the user from creating new addresses. The only concern is for initial synchronization, during which `keypool=<n>` config option for `bitcoind` can be used.

**Comment by JoeGruffins** [comment]
I can do this unless closing, or someone is working on it?

**Comment by chappjc** [comment]
Fair game, @JoeGruffins, thanks.  I think the start to this is to characterize the places we're grabbing addresses that are likely to go unused.  I think the refund addresses are the biggest offender, but we should characterize each spot.  How to resolve depends on the purpose of the address I believe.

**Comment by JoeGruffins** [comment]
It looks to me like all change addresses are used immediately, and therefore not a problem.

The two(ish) places where external addresses are used are in the swap contract and deposit addresses.

In the swap contract we have the address counter-party pays to. I believe once a swap commences, either this address, or the revoke address which should come after that one, and is also external, must be used, so swaps that are not canceled or server revoked are probably never a problem.

However, for orders that are canceled, or server revoked, that address will become unused, and perhaps can be used again right away.

Also, of course, deposit addresses may or may not be used. Maybe it is ok to reuse these after some amount of time has passed. Maybe one day?

I think a good solution to this problem is to focus on those two areas. We could go with number 3 and 4, and add an address database for possibly reusable addresses. With the address, save the time added, an immediately usable field, and a time when we think the address may be used by. 

When a order is canceled, first check that the address isn't already in the database, and if not add it to the database with the "immediately usable" field checked. If already there, check "immediately usable".

When handing out a new deposit address, add it to the database with a "used by" field which is the time 24 hrs from now.

Whenever retrieving a new external address for anything, first look in the db. Check addresses from oldest to newest. If "immediately usable", use and uncheck "immediately usable". Set the "use by" time to 24 hrs from now. For other addresses, if no yet reached "used by" time, pass over. When an address is pass its "use by" time, use `listreceivedbyaddress` to check if it has been used. If yes delete. If no use and reset the "use by" time to 24 hrs from now.
 
If there are no more usable addresses in the db, then fetch a new one. 

Although someone could still recklessly spam the deposit address button, after a day passes they will be re-offered their unused addresses and this will mitigate the problem of restoring from seed and the gap limit.

Hopefully that is all legible?

Also, for every address we must check if it belongs to the current wallet before re-using, or even wipe all saved addresses when a wallet is changed, although I don't think we have the ability to discern if a wallet has been changed yet.

**Comment by JoeGruffins** [comment]
> I think the refund addresses are the biggest offender

For some reason I misread this. We can treat the refund address for successful trades similarly to cancel order's contract address. Whatever the unused address, refund or redeemed, we can reuse immediately I believe.

**Comment by JoeGruffins** [comment]
Ah, bolt, but the idea is still the same.

**Comment by JoeGruffins** [comment]
Another thought, fwiw, if we do reused the unused refund or contract address from an initiated swap, I believe that could create an on-chain link between swaps, as the addresses, although unused, will have been seen by the network. I do not know if this is a privacy concern.

**Comment by JoeGruffins** [comment]
tldr: imo, atm, it's not practical to try and keep our addresses within a gap limit. It makes more sense to make sure hd wallets are synced up to where they need to be to see all their transactions, and ignore the gap limit.

**Comment by ukane-philemon** [comment]
@JoeGruffins, is this still an issue?
